File: Day 3/algorithms/palindrome/palindrome.js
// * Check if a word is a palindrome

// * I need a function that takes in a string
// * I came to the conclusion that I just need to REVERSE the input string
// * and check if its equal to the original input string

function isPalindrome(input) {
  // * Okay, well I can try and loop through the string
  // for (let i = 0; i < input.length; i++)
  // ? This won't work because we need to go in REVERSE order

  // ? Okay, how do I go in reverse?
  // ? Luckily, JavaScript's for loops allow us do directly manipulate the `i` variable at each step

  // * So, firstly, we need to figure out the INITIAL STEP
  // ? `i` has to start from the LAST INDEX of the string
  // ? This means that `i` needs to have the value of `input.length - 1` - as we discussed in the Python example
  // for (let i = input.length - 1; ...)

  // * Okay, then we need to define the ENDING CONDITION
  // ? The loop needs to go on as long as `i` is LARGER THAN OR EQUAL TO 0 (we need `i` to go from some number down to 0)
  // for (let i = input.length - 1; i >= 0; ...)

  // * Lastly, instead of incrementing `i` we just decrement it by 1, for it to go down at each iteration
  // for (let i = input.length - 1; i >= 0; i--)

  // ? We have our for loop, now we create the helper variable for the reversed string
  let reversedInput = "";

  // ? Now we just loop through the input using the for loop we created, and construct the reversed input
  for (let i = input.length - 1; i >= 0; i--) {
    reversedInput += input[i];
  }

  // * Lastly we just check if the two strings are the same, if they are - we found a palindrome!
  if (input === reversedInput) {
    return true;
  } else {
    return false;
  }
}

let result1 = isPalindrome("daniel");
console.log(result1);

let result2 = isPalindrome("kayak");
console.log(result2);

File: Day 3/algorithms/palindrome/palindrome.py
# Check if a word is a palindrome

# * I need a function that takes in a string
# * I came to the conclusion that I just need to REVERSE the input string
# * and check if its equal to the original input string

def is_palindrome(input):
    # * Okay, well I can try and loop through the string
    # for letter in input: # ! This won't work because we need to go in REVERSE order

    # ! Okay, how do I go in reverse?
    # ! Well, what does going in reverse really mean in the context of code?
    # ! It means that we ACCESS elements from the LAST INDEX first and go DOWN TO 0

    # * We can use Python's range() function, to generate numbers in a specific order!
    # ! Okay, well what do we put as that function's arguments?
    # ! Well the first argument (the starting number) needs to be the LAST INDEX of the string, but how do I know what that is?

    # * Python has another function called len() that gives us the length of a string or an array
    # ! If we use len() we can get the length of the string
    # ! Then if we subtract 1 - because Python is 0-indexed - we get the last index of the input string
    input_length = len(input)

    last_index = input_length - 1

    # ! Okay, progress! Now we can use range() to generate numbers from the `last_index` to 0
    # ! Note: The second argument is the ending number, BUT it is EXCLUSIVE, meaning it is not included in the sequence
    # ! Note: This means that instead of a 0, we need to write -1
    # ! Note: range() accepts a third argument, which is the STEP - the difference between each number in the sequence

    # * So this means that in order for us to get the desired sequence, we need to have:
    # * First argument  ->  last_index
    # * Second argument ->  -1
    # * Step            ->  -1
    # for i in range(last_index, -1, -1):

    # * We have the for loop, now we just need a new string variable to store the reverse string in
    reversed_input = ''

    # ! Now we just loop through the input using the for loop we created, and construct the reversed input
    for i in range(last_index, -1, -1):
        reversed_input += input[i]

    # * Lastly we just check if the two strings are the same, if they are - we found a palindrome!
    if input == reversed_input:
        return True
    else:
        return False


result = is_palindrome('kayak')
print(result)
